### PR TITLE
fix issue #13665: Improve index selection when there are multiple candidate indexes

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Fix issue #13665: Improve index selection when there are multiple candidate indexes

This PR improves index selection when there are multiple candidate indexes with common prefixes.

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] Backports required for: 3.7

#### Related Information

- [x] GitHub issue / Jira ticket number: https://github.com/arangodb/arangodb/issues/13665

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [x] This change is already covered by existing tests, such as *aql-projections*.
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new **integration tests** (aql-index-choice.js)

### CODE REVIEW

File: arangod/Indexes/SortedIndexAttributeMatcher.cpp [hunk 1]
```diff
@@ -388,19 +417,11 @@ Index::SortCosts SortedIndexAttributeMatcher::supportsSortCondition(
       if (costs.coveredAttributes >= sortCondition->numAttributes()) {
         // sort is fully covered by index. no additional sort costs!
         // forward iteration does not have high costs
-        costs.estimatedCosts = itemsInIndex * 0.001;
-        if (idx->isPersistent() && sortCondition->isDescending()) {
-          // reverse iteration has higher costs than forward iteration
```
Comment (mpoeter) [inline arangod/Indexes/SortedIndexAttributeMatcher.cpp:393]: Why was this removed? Is this statement no longer true?

File: tests/js/server/aql/aql-index-choice.js [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+        let nodes = AQL_EXPLAIN(q).plan.nodes;
+        assertEqual(1, nodes.filter((n) => n.type === 'IndexNode').length);
+        assertEqual(0, nodes.filter((n) => n.type === 'SortNode').length);
+        let indexNode = nodes.filter((n) => n.type === 'IndexNode')[0];
+        assertEqual([], indexNode.projections);
+        assertFalse(indexNode.indexCoversProjections);
+        assertEqual(1, indexNode.indexes.length);
+        let index = indexNode.indexes[0];
+        assertEqual("persistent", index.type);
+        assertEqual(["uid", "dt"], index.fields);
```
Comment (mpoeter) [inline tests/js/server/aql/aql-index-choice.js:131]: Why would be pick the index with two fields in this scenario? `dt` is not used, neither in a filter nor in a projection.

File: tests/js/server/aql/aql-index-choice.js [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+        let nodes = AQL_EXPLAIN(q).plan.nodes;
+        assertEqual(1, nodes.filter((n) => n.type === 'IndexNode').length);
+        assertEqual(0, nodes.filter((n) => n.type === 'SortNode').length);
+        let indexNode = nodes.filter((n) => n.type === 'IndexNode')[0];
+        assertEqual(["uid"], indexNode.projections);
+        assertTrue(indexNode.indexCoversProjections);
+        assertEqual(1, indexNode.indexes.length);
+        let index = indexNode.indexes[0];
+        assertEqual("persistent", index.type);
+        assertEqual(["uid", "dt"], index.fields);
```
Comment (mpoeter) [inline tests/js/server/aql/aql-index-choice.js:153]: Why would be pick the index with two fields in this scenario? `dt` is not used, neither in a filter nor in a projection.

File: tests/js/server/aql/aql-index-choice.js [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+      db[cn].ensureIndex({ type: "persistent", fields: ["uid"] });
+      
+      [
+        `FOR doc IN ${cn} FILTER doc.uid == 1234 RETURN doc`,
+        `FOR doc IN ${cn} FILTER doc.uid > 1234 RETURN doc`,
+        `FOR doc IN ${cn} FILTER doc.uid >= 1234 RETURN doc`
+      ].forEach((q) => {
+        let nodes = AQL_EXPLAIN(q).plan.nodes;
+        assertEqual(1, nodes.filter((n) => n.type === 'IndexNode').length);
+        assertEqual(0, nodes.filter((n) => n.type === 'SortNode').length);
```
Comment (mpoeter) [inline tests/js/server/aql/aql-index-choice.js:81]: Why the check that no SortNode is found?

File: tests/js/server/aql/aql-index-choice.js [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+        assertEqual(1, nodes.filter((n) => n.type === 'SortNode').length);
+        let indexNode = nodes.filter((n) => n.type === 'IndexNode')[0];
+        assertEqual([], indexNode.projections);
+        assertFalse(indexNode.indexCoversProjections);
+        assertEqual(1, indexNode.indexes.length);
+        let index = indexNode.indexes[0];
+        assertEqual("persistent", index.type);
+        assertEqual(["uid", "dt"], index.fields);
+      });
+    },
```
Comment (mpoeter) [inline tests/js/server/aql/aql-index-choice.js:425]: How about another test case where we filter on `dt` and sort on `uid`? That would be covered by the index, so we should not have a SortNode.

File: tests/js/server/aql/aql-index-choice.js [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+  };
+}
```
Comment (mpoeter) [inline tests/js/server/aql/aql-index-choice.js:468]: How about a test where we have two indexes, one one `uid` and one on `dt`; in the query we filter on `uid` and sort on `dt` or vice versa. Which filter should be picked? The one that covers the filter or the one that covers the sort?

File: arangod/Indexes/SortedIndexAttributeMatcher.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Indexes/SortedIndexAttributeMatcher.cpp:393]: Added this back.

File: tests/js/server/aql/aql-index-choice.js (hunk 2 above)
Comment (jsteemann) [reply tests/js/server/aql/aql-index-choice.js:131]: When all other things are equal, we will always pick the index with more attributes.

File: tests/js/server/aql/aql-index-choice.js (hunk 4 above)
Comment (jsteemann) [reply tests/js/server/aql/aql-index-choice.js]: Removed this superfluous check :+1:

File: tests/js/server/aql/aql-index-choice.js (hunk 5 above)
Comment (jsteemann) [reply tests/js/server/aql/aql-index-choice.js:423]: Added

File: tests/js/server/aql/aql-index-choice.js (hunk 6 above)
Comment (jsteemann) [reply tests/js/server/aql/aql-index-choice.js:521]: Added

**Comment by mpoeter** [comment]
How about adding a few more tests with `COLLECT` instead of `SORT`? With the index the optimizer should always pick the sorted executor.

**Comment by jsteemann** [comment]
@mpoeter : ok, will add such tests!
